rfc-108-including-gem-component-assets.md

```diff
@@ -0,0 +1,189 @@
+
+We now have a number of private applications that use the components gem and components are being written for those applications. While the past model of putting components into applications has worked, we now have components that need to be shared across more than one private application, which is not possible with the current model without adding unnecessary page weight to www.gov.uk.
+
+Additionally, the way that the Sass is currently included in each application makes caching ineffective. When a user lands on a page on GOV.UK, their browser downloads the CSS for that application (say `government-frontend`). This is a single file that includes custom styles from that application plus the styles from the gem. If they then navigate to a page served by another application (say `collections`) they download a new CSS file that contains the exact same styles from the gem, but because it is mixed in with the custom styles for that application, they are different files and cannot be cached.
+
+This problem is potentially limited by gzipping but we could do with some numbers on this.
+
+## Possible solution 1
+
+Instead of including all of the component Sass, each application should be configured to only include the components that it needs. For example, in an applications `_application.scss`:
+
+```
+// currently
+@import 'govuk_publishing_components/all_components';
+```
+
+```
+// proposed
+@import 'govuk_publishing_components/components/accordion';
+@import 'govuk_publishing_components/components/button';
+// (etc.)
+```
+
+This would mean we can add any components we like to the gem without worrying that they are needlessly adding page weight to the publically facing GOV.UK.
+
+## Javascript
+
+We can do something similar with Javascript, however the initialisation of specific modules may be more complicated and require more thought. Additionally knowing whether or not to include jQuery could be difficult.
+
+As a MVP Javascript could be left as it is for now.
+
+### Finding which components are in use
+
+Adding a not-in-use component into an application with this new model would be a relatively safe procedure, as the developer would notice immediately if the styles and Javascript for that component had not been included in the application. However for the future it would be useful if we had some kind of tool that helped us to avoid leaving components unstyled or non-functional.
+
+In the first instance we could search through the code base and write a test for that application to ensure the right assets are being included. This test would need to be kept up to date as components are added and removed.
+
+We could also look at altering the component guide to either mark components in use or put them in a separate list. That way we could look at the component guide for an application and immediately know which components are in use and which are not.
+
+### Certainty that a component renders correctly in isolation
```

Comment:
This is already done in `govuk-frontend` we could use a similar model.

Reply:
`govuk-frontend` uses an iframe to render components but appears to include all of the components CSS in that iframe. That does give some isolation but I think we're looking for the kind of isolation where a single component works by itself.

I've attempted this and there's some problems, better to discuss offline.